File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
    
    int helper(vector<int>& arr, int mid)
{
	int ans=0;
	for(int i=0;i<arr.size();i++)
	{
		ans += ceil(double(arr[i])/mid);
	}
	return ans;
}


    
    
    int smallestDivisor(vector<int>& arr, int limit) {
        	long long s = 1.0;
	long long e = *max_element(arr.begin(),arr.end());
	// cout<<e<<"high"<<endl;
	long long ans = INT_MAX;

	while(s<=e)
	{
		long long mid = s + (e-s)/2;
		int result  = helper(arr, mid);
		
		if(result <=limit)
		{
			ans = min(ans,mid);
			e = mid-1;
		}
		else s = mid+1;
	}
	
	return ans;
    }
};

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.java
class Solution {
     public static int max(int[] b)
    {
        int m = 0;
        for(int i=0;i<b.length;i++)
        {
            if(b[i]>m) m = b[i];
        }
        return m;
    }
    
     public static int helper(int[] b , int mid)
    {
        int temp = 0;
        int count = 0;
        for(int i=0;i<b.length;i++)
        {
            if(b[i] < mid) count ++;
            else if(b[i] == mid) count += b[i]/mid;
            else count += (b[i] + mid -1)/mid; 
        }
        
        return count;
    }
    
    public int smallestDivisor(int[] b, int m) {
         int s = 1;
        int e = max(b);
        
        int ans = Integer.MAX_VALUE;
        
        while(s<=e)
        {
            int mid = s + (e-s)/2;
            int h = helper(b,mid);
            System.out.println(h);
            if(h <= m){
                ans = Math.min(ans, mid);
                e = mid-1;
            }
            else  s = mid+1;
        }
        return ans;
    }
}